### ISSUE TITLE
Problem sending data from 2 slaves to the master.

### ISSUE THREAD

**Issue by pauPeixo** [tonton81/SPI_MSTransfer_T4#9]
I am trying to send data from 2 Teensy 4.1 slaves to the master (Teensy 4.1) , but I am not receiving the data at the master. When there is only one slave, it works perfectly, but when there are 2 slaves, it doesn't work.

The **master** code is:
`#include "SPI.h"
#include "SPI_MSTransfer_MASTER.h"
SPI_MSTransfer_MASTER<&SPI, 10, 0x1234> mySPI1234;
SPI_MSTransfer_MASTER<&SPI, 10, 0x1235> mySPI4567;

void setup() {
  Serial.begin(115200);
  SPI.begin();
  pinMode(10, OUTPUT);
  digitalWrite(10, 1);
  mySPI1234.begin();  
  mySPI1234.onTransfer(myCB);  
  mySPI4567.begin();  
  mySPI4567.onTransfer(myCB);  
}

void loop() {
  mySPI1234.events();
  mySPI4567.events();
  static uint32_t t = millis();
  if ( millis() - t > 1000 ) {
    Serial.println(millis());
    uint16_t buf[5] = { 2, 0xF2, 0xDEAD, 0xF4, 0xBEEF };
    uint16_t buf2[5] = { 1, 0xF7, 0xF8, 0xF9, 0xDEAD };
    mySPI1234.transfer16(buf2, 5, random(0x0000, 0xFFFF));
    mySPI4567.transfer16(buf, 5, random(0x0000, 0xFFFF));
    t = millis();
  }
}
void myCB(uint16_t *buffer, uint16_t length, AsyncMST info) {
  for ( int i = 0; i < length; i++ ) {
    Serial.print(buffer[i], HEX); Serial.print(" ");
  }
  Serial.print(" --> Length: "); Serial.print(length);
  Serial.print(" --> PacketID: "); Serial.println(info.packetID, HEX);
}`
** Slave 1 **

`#include "SPI_MSTransfer_T4.h"
SPI_MSTransfer_T4<&SPI, 0x1234> mySPI;

void setup() {
  Serial.begin(115200);
  mySPI.begin();
  mySPI.onTransfer(myCB);
}

void loop() {
  mySPI.events();
  static uint32_t t = millis();
  if ( millis() - t > 1000 ) {
    Serial.print("millis: "); Serial.println(millis());
    t = millis();
    uint16_t buf2[5] = { 4, 0xF7, 0xF8, 0xF9, 0xDEAD };
    mySPI.transfer16(buf2, 5, random(0x0000, 0xFFFF));
  }
}

void myCB(uint16_t *buffer, uint16_t length, AsyncMST info) {
  for ( int i = 0; i < length; i++ ) {
    Serial.print(buffer[i], HEX); Serial.print(" ");
  }
  Serial.print(" --> Length: "); Serial.print(length);
  Serial.print(" --> PacketID: "); Serial.println(info.packetID, HEX);
}`

**Slave 2:** 
`#include "SPI_MSTransfer_T4.h"
SPI_MSTransfer_T4<&SPI, 0x1235> mySPI;

void setup() {
  Serial.begin(115200);
  mySPI.begin();
  mySPI.onTransfer(myCB);
}

void loop() {
  mySPI.events();
  static uint32_t t = millis();
  if ( millis() - t > 500 ) {
    Serial.print("millis: "); Serial.println(millis());
    t = millis();
    uint16_t buf2[5] = { 5, 0xF7, 0xF8, 0xF9, 0xDEAD };
    mySPI.transfer16(buf2, 5, random(0x0000, 0xFFFF));
  }
}

void myCB(uint16_t *buffer, uint16_t length, AsyncMST info) {
  for ( int i = 0; i < length; i++ ) {
    Serial.print(buffer[i], HEX); Serial.print(" ");
  }
  Serial.print(" --> Length: "); Serial.print(length);
  Serial.print(" --> PacketID: "); Serial.println(info.packetID, HEX);
}`

I have thought that it might be a problem with the CS (Chip Select). I want to change the CS pin of the slave, but I don't know how to change it. Does anyone know how to determine which pin is the CS on the slave?

I have discovered that if I comment out line 112 in the library, which designates the pin for MOSI only in Slave 2, then the master does receive data from Slave 1 but not from Slave 2.

IOMUXC_SW_MUX_CTL_PAD_GPIO_B0_02=0x03;

**Comment by KR-0011** [comment]
Hey,

I am facing the same issue while communicating with 2 slaves. My implementation seems to be somewhat similar to yours. Could you kindly let me know if you were able to get around with it?

**Comment by pauPeixo** [comment]
Yes I solved it, but it was not a problem with the code. I cannot use a different CS for the Slaves. The CS for the slaves is pin 10 and cannot be changed. In the master code you can select which pin to use for each CS. What is not possible is to share 2 slave devices on the same CS.

**Comment by KR-0011** [comment]
Thank you very much for your response. Could you kindly let me know how was it possible to connect the same MISO, MOSI and SCK lines across all the three devices. I was unable to see any response when I tried to connect them using a breadboard.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
